File: src/util/NMIvalidator.py
class NMIValidator:

    """
    Class to assess if supplied NMI's are valid, and append a checksum if necessary.
    """

    def __init__(self, _NMI):

        self.is_valid = False
        self.NMI = self.validate_NMI(_NMI)
        self.checksum = self.calculate_checksum(_NMI)
        # print(f"validated NMI is {self.NMI}")

    def validate_NMI(self, NMI):

        if(NMI.isdigit() == False):
            # print("NMI must be digits only...")
            return None

        length = len(NMI)
        # print(f"NMI len is {length}")
        
        if(length < 10 or length > 11):
            # print("Invalid NMI (should be 10 or 11 digits)...")
            return None

        if(length == 11):
            # print("Checksum is valid")
            self.is_valid = True
            return NMI

        else:
            # print("Appended checksum")
            self.is_valid = True
            NMI = NMI + self.calculate_checksum(NMI)
            return NMI

    def sum_of_list(self, l):
        """Helper function to calculate sum of list of integers"""

        total = 0
        for val in l:
            total = total + val
        return total

    def calculate_checksum(self, NMI):
    
        """
        The checksum formula is: 
        1. Reverse the number order then take the ASCII value of each digit
        2. Double every second ASCII value
        3. Sum the digits making up each value (eg. 48 becomes 4+8=12)
        4. Sum all of these values together
        5. Round this total up to the nearest 10
        6. The amount required to round up to 10 is the checksum, eg. if total is 72 the checksum is 8
        """
        
        NMI_reversed = NMI[::-1]
        
        ASCII_values = []
            
        for character in NMI_reversed:
            ASCII_values.append(ord(character))
            
        # Double every second ASCII value
        double_ASCII = [n if i%2 else 2*n for i,n in enumerate(ASCII_values)]
        
        sum_digit = []
        for ele in double_ASCII:
            sum = 0
            for digit in str(ele):
                sum += int(digit)
            sum_digit.append(sum)

        list_sum = self.sum_of_list(sum_digit)
        
        # Round list_sum to the nearest 10 and find the distance
        check_sum = round(list_sum,-1) - list_sum
        
        if check_sum < 0:
            check_sum = check_sum + 10

        self.checksum = check_sum
        return str(check_sum)

File: src/util/NMIvalidator.js
function NMIvalidation(NMI) {
  NMI = NMI.replace(/\s/g, "");

  //Check if length is 10 or 11. If 10 return True, else execute checksum
  if (NMI.length === 10) {
    return true;
  } else if (NMI.length === 11) {
    return checkSumValidation(NMI);
  }

  return false;
}

function checkSumValidation(NMI) {
  const lastDigit = NMI.charAt(NMI.length - 1);
  const NMI_reversed = NMI.slice(0, -1).split("").reverse();

  let ASCII_values = [];
  let double_ASCII = [];

  for (let index = 0; index < NMI_reversed.length; index++) {
    const element = NMI_reversed[index];
    ASCII_values.push(element.charCodeAt(0));
  }

  for (let index = 0; index < ASCII_values.length; index++) {
    const element = ASCII_values[index];
    if (index % 2 === 0) {
      double_ASCII.push(element * 2);
    } else {
      double_ASCII.push(element);
    }
  }

  const sum_digit = [];
  for (let index = 0; index < double_ASCII.length; index++) {
    let sum = 0;
    const element = double_ASCII[index].toString();
    for (let index = 0; index < element.length; index++) {
      const e = element[index];
      sum += parseInt(e);
    }
    sum_digit.push(sum);
  }

  const list_sum = sum_digit.reduce((partialSum, a) => partialSum + a, 0);

  let check_sum = Math.ceil(list_sum / 10) * 10 - list_sum;

  if (check_sum < 0) {
    check_sum = check_sum + 10;
  }

  if (check_sum === Number(lastDigit)) {
    return true;
  }
  return false;
}

export default NMIvalidation;
